### PR TITLE
Temporary Table support in unsharded keyspace

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This adds temp table support to unsharded keyspace.

## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/5371, https://github.com/vitessio/vitess/issues/7407

## Checklist
- [x] Tests were added
- [x] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

-  Query Serving

### CODE REVIEW

**Comment by systay** [comment]
@harshit-gangal we need to document the caching behaviour

**Comment by harshit-gangal** [comment]
We need to document temp table support. :)

> @harshit-gangal we need to document the caching behaviour

**Comment by harshit-gangal** [comment]
Documentation: https://github.com/vitessio/website/pull/702

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -3170,11 +3213,17 @@ func (node *AlterDatabase) Format(buf *TrackedBuffer) {
 
 // Format formats the node.
 func (node *CreateTable) Format(buf *TrackedBuffer) {
+	buf.WriteString("create ")
+	if node.Temp {
+		buf.WriteString("temporary ")
+	}
+	buf.WriteString("table ")
+
```
Comment (shlomi-noach) [inline go/vt/sqlparser/ast.go:3221]: :+1:

File: go/vt/sqlparser/ast.go [hunk 2]
```diff
@@ -677,6 +680,46 @@ func (node *AlterView) IsFullyParsed() bool {
...
+// IsTemporary implements the DDLStatement interface
+func (node *DropView) IsTemporary() bool {
+	return false
+}
+
+// IsTemporary implements the DDLStatement interface
+func (node *DropTable) IsTemporary() bool {
+	return node.Temp
+}
```
Comment (shlomi-noach) [inline go/vt/sqlparser/ast.go:716]: :+1:

File: go/vt/sqlparser/sql.y [hunk 3]
```diff
@@ -827,9 +827,9 @@ vindex_param:
   }
 
 create_table_prefix:
-  CREATE TABLE not_exists_opt table_name
+  CREATE temp_opt TABLE not_exists_opt table_name
   {
-    $$ = &CreateTable{Table: $4, IfNotExists: $3}
+    $$ = &CreateTable{Table: $5, IfNotExists: $4, Temp: $2}
```
Comment (shlomi-noach) [inline go/vt/sqlparser/sql.y:832]: :+1:

File: go/vt/vtgate/engine/ddl.go [hunk 4]
```diff
@@ -37,71 +37,83 @@ type DDL struct {
...
 	return false
 }
 
 // Execute implements the Primitive interface
-func (v *DDL) Execute(vcursor VCursor, bindVars map[string]*query.BindVariable, wantfields bool) (result *sqltypes.Result, err error) {
+func (ddl *DDL) Execute(vcursor VCursor, bindVars map[string]*query.BindVariable, wantfields bool) (result *sqltypes.Result, err error) {
+	if ddl.CreateTempTable {
+		vcursor.Session().HasCreatedTempTable()
+		vcursor.Session().NeedsReservedConn()
+		return ddl.NormalDDL.Execute(vcursor, bindVars, wantfields)
```
Comment (shlomi-noach) [inline go/vt/vtgate/engine/ddl.go:90]: :+1:

File: go/vt/vtgate/planbuilder/ddl.go [hunk 5]
```diff
@@ -31,12 +31,20 @@ func buildGeneralDDLPlan(sql string, ddlStatement sqlparser.DDLStatement, vschem
 		return nil, err
 	}
 
+	if ddlStatement.IsTemporary() {
+		if normalDDLPlan.Keyspace.Sharded {
+			return nil, vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "Temporary table not supported in sharded keyspace: %s", normalDDLPlan.Keyspace.Name)
+		}
+		onlineDDLPlan = nil // emptying this so it does not accidentally gets used somewhere
```
Comment (shlomi-noach) [inline go/vt/vtgate/planbuilder/ddl.go:38]: :+1:

File: go/vt/vtgate/planbuilder/ddl.go [hunk 6]
```diff
@@ -31,12 +31,20 @@ func buildGeneralDDLPlan(sql string, ddlStatement sqlparser.DDLStatement, vschem
...
-		SQL:       normalDDLPlan.Query,
-		DDL:       ddlStatement,
-		NormalDDL: normalDDLPlan,
-		OnlineDDL: onlineDDLPlan,
+		Keyspace:        normalDDLPlan.Keyspace,
+		SQL:             normalDDLPlan.Query,
+		DDL:             ddlStatement,
+		NormalDDL:       normalDDLPlan,
+		OnlineDDL:       onlineDDLPlan,
+		CreateTempTable: ddlStatement.IsTemporary(),
```
Comment (shlomi-noach) [inline go/vt/vtgate/planbuilder/ddl.go:47]: It feels like a bit of high level location to place this low level information, but I see why this is necessary.

**Review by shlomi-noach — APPROVED** [review]
:racing_car:

**Pull request merged**